File: scripts/qa-content.js
#!/usr/bin/env node
/**
 * QA Content Script — Baitul Hikmah
 * 
 * Checks all event content files for quality issues:
 * - Bibliography: duplicates, too many entries, orphan citations
 * - Citations: missing refs, citation 0, gaps
 * - Arabic text: used outside Quran/Hadits
 * - Structure: missing frontmatter, wrong section order
 * 
 * Usage: node scripts/qa-content.js [event-dir]
 *   No args = check all events
 *   With arg = check specific event (e.g. e03-perjalanan-syam)
 */

const fs = require('fs')
const path = require('path')

const CONTENT_DIR = path.join(__dirname, '..', '..', '..', 'docs', 'content', 'events')

// Unicode superscript → number
const supMap = { '⁰': 0, '¹': 1, '²': 2, '³': 3, '⁴': 4, '⁵': 5, '⁶': 6, '⁷': 7, '⁸': 8, '⁹': 9 }
function parseSuperscripts(text) {
  const nums = new Set()
  // Match unicode superscripts (single or multi-digit like ¹² ²⁶)
  const supRegex = /[⁰¹²³⁴⁵⁶⁷⁸⁹]+/g
  let m
  while ((m = supRegex.exec(text)) !== null) {
    const digits = m[0].split('').map(c => supMap[c]).join('')
    nums.add(parseInt(digits, 10))
  }
  // Match ^N style
  const caretRegex = /\^(\d+)/g
  while ((m = caretRegex.exec(text)) !== null) {
    nums.add(parseInt(m[1], 10))
  }
  return nums
}

function countBibEntries(text) {
  const bibSection = text.split(/^## Daftar Pustaka|^## Bibliography/m)[1]
  if (!bibSection) return { count: 0, entries: [], duplicates: [] }
  
  const entries = []
  const lines = bibSection.split('\n')
  for (const line of lines) {
    const match = line.match(/^\d+\.\s*(.+)/)
    if (match) entries.push(match[1].trim())
  }
  
  // Detect duplicates by normalizing
  const normalize = (s) => s.toLowerCase().replace(/[,.*\s]+/g, ' ').replace(/\s+/g, ' ').trim()
  const seen = new Map()
  const duplicates = []
  for (let i = 0; i < entries.length; i++) {
    // Extract main source (author + book title, ignore bab/page details)
    const key = normalize(entries[i]).replace(/\bjilid \d+\b/g, '').replace(/\bbab .+$/g, '').replace(/\bhlm\b.+$/g, '').trim()
    if (seen.has(key)) {
      duplicates.push({ idx: i + 1, entry: entries[i], duplicateOf: seen.get(key) + 1 })
    } else {
      seen.set(key, i)
    }
  }
  
  return { count: entries.length, entries, duplicates }
}

function checkFile(filePath, fileName) {
  const issues = []
  const text = fs.readFileSync(filePath, 'utf8')
  
  // Check frontmatter
  if (!text.startsWith('---')) {
    issues.push({ severity: 'ERROR', msg: 'Missing frontmatter' })
  }
  
  // Strip frontmatter for content analysis
  const content = text.replace(/^---[\s\S]*?---\n*/, '')
  
  if (fileName.startsWith('general-')) {
    // === GENERAL MODE CHECKS ===
    
    // Bibliography checks
    const bib = countBibEntries(content)
    if (bib.count === 0) {
      issues.push({ severity: 'ERROR', msg: 'No Daftar Pustaka found' })
    } else {
      if (bib.count > 10) {
        issues.push({ severity: 'ERROR', msg: `Too many bibliography entries: ${bib.count} (max 8-10). Likely footnote-style instead of consolidated.` })
      } else if (bib.count > 8) {
        issues.push({ severity: 'WARN', msg: `Bibliography has ${bib.count} entries (recommended max 8)` })
      }
      
      if (bib.duplicates.length > 0) {
        for (const d of bib.duplicates) {
          issues.push({ severity: 'ERROR', msg: `Duplicate bib entry #${d.idx} ≈ #${d.duplicateOf}: "${d.entry.slice(0, 60)}..."` })
        }
      }
    }
    
    // Citation checks
    const bodyBeforeBib = content.split(/^## Daftar Pustaka|^## Bibliography/m)[0] || content
    const citedNums = parseSuperscripts(bodyBeforeBib)
    
    if (citedNums.has(0)) {
      issues.push({ severity: 'ERROR', msg: 'Citation ⁰ (zero) found — citations must start from 1' })
    }
    
    if (bib.count > 0) {
      // Check for orphan citations (cited but not in bib)
      for (const n of citedNums) {
        if (n > bib.count) {
          issues.push({ severity: 'ERROR', msg: `Citation ${n} referenced but only ${bib.count} bib entries` })
        }
      }
      
      // Check for unused bib entries
      for (let i = 1; i <= bib.count; i++) {
        if (!citedNums.has(i)) {
          issues.push({ severity: 'WARN', msg: `Bib entry #${i} never cited in text` })
        }
      }
    }
    
    if (citedNums.size === 0) {
      issues.push({ severity: 'WARN', msg: 'No superscript citations found in text' })
    }
    
    // Check structure
    if (!content.includes('## ')) {
      issues.push({ severity: 'ERROR', msg: 'No h2 sections found' })
    }
  }
  
  if (fileName.startsWith('children-')) {
    // === CHILDREN MODE CHECKS ===
    
    const sections = content.split(/^## /m).filter(s => s.trim())
    // Subtract frontmatter section if present
    const realSections = sections.filter(s => !s.startsWith('title:') && !s.startsWith('---'))
    
    if (realSections.length < 3) {
      issues.push({ severity: 'WARN', msg: `Only ${realSections.length} sections (expected 4+)` })
    }
    
    // Check for illustration briefs
    const briefCount = (content.match(/🎨/g) || []).length
    if (briefCount === 0) {
      issues.push({ severity: 'WARN', msg: 'No 🎨 illustration briefs found' })
    }
    
    // Check for tables (not allowed in children mode)
    if (content.includes('|---') || content.includes('| ---')) {
      issues.push({ severity: 'ERROR', msg: 'Table found in children mode (not allowed)' })
    }
  }
  
  // === COMMON CHECKS ===
  
  // Check ID ↔ EN section count consistency would need both files
  
  return issues
}

function checkEvent(eventDir) {
  const results = { dir: path.basename(eventDir), files: {} }
  const files = fs.readdirSync(eventDir).filter(f => f.endsWith('.md'))
  
  for (const file of files) {
    const issues = checkFile(path.join(eventDir, file), file)
    if (issues.length > 0) {
      results.files[file] = issues
    }
  }
  
  // Cross-file checks: ID ↔ EN section count
  for (const lang of ['id', 'en']) {
    const genFile = `general-${lang}.md`
    const childFile = `children-${lang}.md`
    // Could add cross-checks here
  }
  
  // Cross-check children-id vs children-en section count
  const childIdPath = path.join(eventDir, 'children-id.md')
  const childEnPath = path.join(eventDir, 'children-en.md')
  if (fs.existsSync(childIdPath) && fs.existsSync(childEnPath)) {
    const idSections = fs.readFileSync(childIdPath, 'utf8').split(/^## /m).length
    const enSections = fs.readFileSync(childEnPath, 'utf8').split(/^## /m).length
    if (idSections !== enSections) {
      if (!results.files['children-id.md']) results.files['children-id.md'] = []
      results.files['children-id.md'].push({
        severity: 'ERROR',
        msg: `Section count mismatch: children-id has ${idSections - 1}, children-en has ${enSections - 1}`
      })
    }
  }
  
  return results
}

// Main
const targetEvent = process.argv[2]
let dirs

if (targetEvent) {
  const fullPath = path.join(CONTENT_DIR, targetEvent)
  if (!fs.existsSync(fullPath)) {
    console.error(`Event dir not found: ${fullPath}`)
    process.exit(1)
  }
  dirs = [fullPath]
} else {
  dirs = fs.readdirSync(CONTENT_DIR)
    .filter(d => d.startsWith('e') && fs.statSync(path.join(CONTENT_DIR, d)).isDirectory())
    .sort()
    .map(d => path.join(CONTENT_DIR, d))
}

let totalErrors = 0
let totalWarns = 0
let cleanEvents = 0
const summary = { errors: [], warns: [] }

for (const dir of dirs) {
  const result = checkEvent(dir)
  const fileKeys = Object.keys(result.files)
  
  if (fileKeys.length === 0) {
    cleanEvents++
    continue
  }
  
  console.log(`\n📁 ${result.dir}`)
  for (const [file, issues] of Object.entries(result.files)) {
    for (const issue of issues) {
      const icon = issue.severity === 'ERROR' ? '❌' : '⚠️'
      console.log(`  ${icon} ${file}: ${issue.msg}`)
      if (issue.severity === 'ERROR') {
        totalErrors++
        summary.errors.push(`${result.dir}/${file}: ${issue.msg}`)
      } else {
        totalWarns++
        summary.warns.push(`${result.dir}/${file}: ${issue.msg}`)
      }
    }
  }
}

console.log(`\n${'═'.repeat(60)}`)
console.log(`📊 QA Summary: ${dirs.length} events checked`)
console.log(`   ✅ Clean: ${cleanEvents}`)
console.log(`   ❌ Errors: ${totalErrors}`)
console.log(`   ⚠️  Warnings: ${totalWarns}`)

if (totalErrors > 0) {
  console.log(`\n❌ TOP ERRORS:`)
  for (const e of summary.errors.slice(0, 20)) {
    console.log(`   ${e}`)
  }
}

process.exit(totalErrors > 0 ? 1 : 0)

File: scripts/qa-content.py
#!/usr/bin/env python3
"""
QA Content Review — Baitul Hikmah
Automated quality gate for content files before deploy.

Usage:
  python3 scripts/qa-content.py                    # Audit all events
  python3 scripts/qa-content.py e03 e47            # Audit specific events
  python3 scripts/qa-content.py --json             # Output as JSON (for automation)

Exit codes:
  0 = all PASS
  1 = some FAIL (details in output)

Violation Types:
  V1: Pustaka tanpa sitasi (entry #N tanpa ^N di body)
  V2: Sitasi tanpa pustaka (^N di body tanpa entry #N)
  V3: Duplikat pustaka (sumber sama muncul >1x)
  V4: Pustaka count too few (<3)
  V5: Sitasi ^0 (nol)
  V6: ID↔EN section count mismatch (general)
  V7: ID↔EN children slide count mismatch
"""

import os, re, sys, json, glob

CONTENT_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "content", "events")
# Fallback to workspace docs if project content dir empty
DOCS_CONTENT_DIR = "/workspace/docs/content/events"


def find_content_dir():
    if os.path.exists(CONTENT_DIR) and os.listdir(CONTENT_DIR):
        return CONTENT_DIR
    return DOCS_CONTENT_DIR


def parse_citations(text):
    """Extract all citation numbers from body text (before bibliography section)."""
    # Split at bibliography
    parts = re.split(r'^## (?:Daftar Pustaka|Referensi|Bibliography)', text, flags=re.MULTILINE)
    if len(parts) < 2:
        return set(), [], []

    body = parts[0]
    bib_section = parts[1]

    # Find caret refs: ^1, ^2, etc.
    caret_refs = set(int(x) for x in re.findall(r'\^(\d+)', body))

    # Find unicode superscripts: ¹²³⁴⁵⁶⁷⁸⁹⁰
    sup_map = {
        '\u00B9': 1, '\u00B2': 2, '\u00B3': 3,
        '\u2074': 4, '\u2075': 5, '\u2076': 6,
        '\u2077': 7, '\u2078': 8, '\u2079': 9, '\u2070': 0
    }
    unicode_refs = set()
    for ch, num in sup_map.items():
        if ch in body:
            unicode_refs.add(num)

    all_refs = caret_refs | unicode_refs

    # Parse bibliography entries
    bib_entries = re.findall(r'^\d+\.\s*(.+)$', bib_section, re.MULTILINE)
    bib_count = len(bib_entries)

    return all_refs, bib_entries, bib_count


def normalize_bib(entry):
    """Normalize bibliography entry for duplicate detection."""
    return re.sub(r'[^a-zA-Z0-9]', '', entry.lower())


def count_sections(text):
    """Count ## headings (excluding Daftar Pustaka)."""
    headings = re.findall(r'^## (.+)$', text, re.MULTILINE)
    return [h for h in headings if not re.match(r'(?:Daftar Pustaka|Referensi|Bibliography|References)', h)]


def count_children_slides(text):
    """Count content sections in children file (split by ## or ---)."""
    sections = re.split(r'^## ', text, flags=re.MULTILINE)
    # Filter empty and frontmatter
    valid = [s for s in sections if s.strip() and not s.strip().startswith('title:')]
    return len(valid)


def audit_event(event_dir):
    """Audit a single event directory. Returns (event_name, pass/fail, violations)."""
    dirname = os.path.basename(event_dir.rstrip('/'))
    violations = []

    # Check general-id.md exists
    gen_id_path = os.path.join(event_dir, "general-id.md")
    if not os.path.exists(gen_id_path):
        return dirname, False, ["MISSING: general-id.md not found"]

    gen_id = open(gen_id_path, 'r').read()
    gen_id_clean = re.sub(r'^---.*?---\n*', '', gen_id, flags=re.DOTALL)

    # === Citation checks (V1-V5) ===
    all_refs, bib_entries, bib_count = parse_citations(gen_id_clean)

    if bib_count == 0:
        violations.append("V4: Tidak ada section Daftar Pustaka")
    else:
        # V4: Minimum count (no max — any amount OK if all cited & consolidated)
        if bib_count < 3:
            violations.append(f"V4: Pustaka terlalu sedikit ({bib_count} < 3)")

        # V1: Pustaka tanpa sitasi
        uncited = [i for i in range(1, bib_count + 1) if i not in all_refs]
        if uncited:
            violations.append(f"V1: Pustaka tanpa sitasi: entry {uncited}")

        # V2: Sitasi tanpa pustaka
        orphans = [r for r in sorted(all_refs) if r > bib_count or r < 1]
        if orphans:
            violations.append(f"V2: Sitasi tanpa pustaka: ^{orphans}")

        # V5: Sitasi nol
        if 0 in all_refs:
            violations.append("V5: Ada sitasi ^0 (nol) — harus mulai dari ^1")

        # V3: Duplikat pustaka
        dupes = []
        for i in range(len(bib_entries)):
            for j in range(i + 1, len(bib_entries)):
                a = normalize_bib(bib_entries[i])
                b = normalize_bib(bib_entries[j])
                if len(a) > 20 and len(b) > 20 and (a == b or a in b or b in a):
                    dupes.append(f"#{i+1} ≈ #{j+1}")
        if dupes:
            violations.append(f"V3: Duplikat pustaka: {dupes[:5]}{'...' if len(dupes) > 5 else ''}")

    # === ID↔EN sync checks (V6-V7) ===
    gen_en_path = os.path.join(event_dir, "general-en.md")
    if os.path.exists(gen_en_path):
        gen_en = open(gen_en_path, 'r').read()
        gen_en_clean = re.sub(r'^---.*?---\n*', '', gen_en, flags=re.DOTALL)
        id_sections = count_sections(gen_id_clean)
        en_sections = count_sections(gen_en_clean)
        if len(id_sections) != len(en_sections):
            violations.append(f"V6: Section count mismatch — general-id has {len(id_sections)}, general-en has {len(en_sections)}")

    child_id_path = os.path.join(event_dir, "children-id.md")
    child_en_path = os.path.join(event_dir, "children-en.md")
    if os.path.exists(child_id_path) and os.path.exists(child_en_path):
        child_id = open(child_id_path, 'r').read()
        child_en = open(child_en_path, 'r').read()
        child_id_clean = re.sub(r'^---.*?---\n*', '', child_id, flags=re.DOTALL)
        child_en_clean = re.sub(r'^---.*?---\n*', '', child_en, flags=re.DOTALL)
        id_slides = count_children_slides(child_id_clean)
        en_slides = count_children_slides(child_en_clean)
        if id_slides != en_slides:
            violations.append(f"V7: Children slide count mismatch — id has {id_slides}, en has {en_slides}")

    # === File completeness ===
    expected_files = ["general-id.md", "general-en.md", "children-id.md", "children-en.md"]
    missing = [f for f in expected_files if not os.path.exists(os.path.join(event_dir, f))]
    if missing:
        violations.append(f"INCOMPLETE: Missing files: {missing}")

    passed = len(violations) == 0
    return dirname, passed, violations


def main():
    content_dir = find_content_dir()
    json_mode = "--json" in sys.argv
    specific = [a for a in sys.argv[1:] if not a.startswith("--")]

    results = {"pass": [], "fail": []}

    event_dirs = sorted(glob.glob(f"{content_dir}/e*/"))

    for event_dir in event_dirs:
        dirname = os.path.basename(event_dir.rstrip('/'))

        # Filter specific events if requested
        if specific:
            event_id = dirname.split('-')[0]
            if event_id not in specific and dirname not in specific:
                continue

        # Skip empty dirs
        if not any(f.endswith('.md') for f in os.listdir(event_dir)):
            continue

        name, passed, violations = audit_event(event_dir)

        if passed:
            results["pass"].append(name)
        else:
            results["fail"].append({"event": name, "violations": violations})

    if json_mode:
        print(json.dumps(results, indent=2, ensure_ascii=False))
    else:
        total = len(results["pass"]) + len(results["fail"])
        print(f"QA CONTENT REVIEW — {total} events audited")
        print(f"✅ PASS: {len(results['pass'])}  |  🔴 FAIL: {len(results['fail'])}")
        print()

        if results["fail"]:
            print("=" * 60)
            print("FAILURES:")
            print("=" * 60)
            for item in results["fail"]:
                print(f"\n🔴 {item['event']}")
                for v in item["violations"]:
                    print(f"   → {v}")

            print()
            print("=" * 60)
            print(f"PASS list ({len(results['pass'])}): {', '.join(results['pass'][:10])}{'...' if len(results['pass']) > 10 else ''}")
        else:
            print("🎉 All events passed QA!")

    sys.exit(1 if results["fail"] else 0)


if __name__ == "__main__":
    main()
